### ISSUE TITLE
Node pbkdf2 defaults to `ascii` encoding?

### ISSUE THREAD

**Issue by dcousens** [crypto-browserify/pbkdf2#11]
Is this intentional or a bug?  Its burning me as the browserify code [now] defaults to `utf8` for strings because it encodes it as a `Buffer` straight away.  Whereas the node seems to be defaulting string inputs to `ascii` encoding.

This therefore gives completely different results to what is expected... sigh.

See https://github.com/crypto-browserify/pbkdf2/commit/ea1ddaf28d06a288a955ae41b717011c1596fb0b for tests.

@calvinmetcalf do you know if this is meant to be the case?  Node bug?

**Comment by dcousens** [comment]
@indutny any ideas?

**Comment by dcousens** [comment]
As can be seen, this is occurring for `sha1` too.

**Comment by calvinmetcalf** [comment]
Is this in just the pure js one or the one that defaults to native?

**Comment by dcousens** [comment]
Native

**Comment by calvinmetcalf** [comment]
the native and js versions parse Unicode salts differently?

**Comment by dcousens** [comment]
@calvinmetcalf did you look at the test?

`require("crypto").pbkdf2Sync`, when giving a `salt`, where `typeof salt === "string"`, appears to treat the `salt` as the following: `new Buffer(salt, "ascii")`. 

This can be evidenced by checking that the result is as *expected* if you manually pass `new Buffer(salt, "utf8")` to `pbkdf2Sync` instead.

**Comment by dcousens** [comment]
@calvinmetcalf did you look at the test?

`require("crypto").pbkdf2Sync` (*native*), when given a `salt`, where `typeof salt === "string"`, appears to treat the `salt` internally as the following: `new Buffer(salt, "ascii")`. 

This can be evidenced by checking that the result is as *expected* if you manually pass `new Buffer(salt, "utf8")` to `pbkdf2Sync` instead.

In short, run the following in node `0.12`:

``` javascript
var salt = "ヘ(^_^ヘ)(ノ^_^)ノ"
var saltBuffer = new Buffer(salt)

var crypto = require("crypto")
var a = crypto.pbkdf2Sync("1234", salt, 1, 16).toString("hex")
var b = crypto.pbkdf2Sync("1234", saltBuffer, 1, 16).toString("hex")

console.log(a, b)
```

**Comment by calvinmetcalf** [comment]
yes the native version does this technically I believe they call it 'binary' not ascii but its due to some legacy stuff in the crypto library which defaults to that other encoding https://github.com/iojs/io.js/blob/2b63bcd247362ec5eb46e235a0eec68d7724a6d7/lib/crypto.js#L26-L35

**Comment by dcousens** [comment]
Should we be replicating that behaviour then?

**Comment by calvinmetcalf** [comment]
yes

**Comment by dcousens** [comment]
Fixed in #12
